compiler/packages/babel-plugin-react-compiler/src/TypeInference/InferTypes.ts

```diff
@@ -777,6 +777,15 @@ class Unifier {
       return {kind: 'Phi', operands: type.operands.map(o => this.get(o))};
     }
 
+    if (type.kind === 'Function') {
+      return {
+        kind: 'Function',
+        isConstructor: type.isConstructor,
+        shapeId: type.shapeId,
+        return: this.get(type.return),
```

Comment:
Seems reasonable, although should we also error on union types that _could_ return JSX? Eg

```js
function Component() {
  const renderItem = item => {
    if (cond) {
      global.property = true;
      return;
    } else {
      return <Item item={item} value={rand} />;
    }
  };
  return <ItemList renderItem={renderItem} />;
}
```

Reply:
done!